bulk-updater/src/operations.rs

```diff
@@ -0,0 +1,304 @@
+use obs_service_cargo::cli::{Compression, Opts, Src, Vendor};
+use std::path::{Path, PathBuf};
+use std::process;
+use std::process::Output;
+use std::{fs, io};
+
+fn do_services(pkgpath: &Path) -> io::Result<()> {
+    let cwd = std::env::current_dir()?;
+    let pkgpath_binding = cwd.join(pkgpath);
+    let cwd_plus_pkgpath = pkgpath_binding.as_os_str().to_string_lossy();
+    let bindmount_option = format!(
+        "{}:{}",
+        cwd.as_os_str().to_string_lossy(),
+        cwd.as_os_str().to_string_lossy()
+    );
+    let cmd_arguments = vec![
+        "--really_quiet",
+        "--config",
+        "scan.cfg",
+        "--cwd",
+        &cwd_plus_pkgpath,
+        "--bindmount",
+        &bindmount_option,
+        "/usr/bin/osc",
+        "service",
+        "ra",
+    ];
+
+    let full_command = process::Command::new("nsjail")
+        .args(cmd_arguments)
+        .current_dir(&cwd)
+        .output()
+        .map_err(|err| {
+            tracing::error!(err = ?err, "Unable to run nsjail");
+            io::Error::new(io::ErrorKind::Interrupted, "Unable to run nsjail")
+        })?;
+    if full_command.status.success() {
+        let command_output = String::from_utf8_lossy(&full_command.stdout);
+        tracing::info!("✅ -- services passed");
+        tracing::info!("stdout -- {}", command_output);
+    } else {
+        let command_output = String::from_utf8_lossy(&full_command.stderr);
+        tracing::info!("🚨 -- services failed");
+        tracing::info!("stderr -- {}", command_output);
+    }
+    Ok(())
+}
+
+pub fn checkout_or_update(pkgname: &str, basepath: &Path) -> io::Result<PathBuf> {
+    let pkgpath = std::path::PathBuf::from(format!("{}:{}", basepath.to_string_lossy(), pkgname));
+    if pkgpath.exists() {
+        tracing::info!("osc revert {}", pkgpath.to_string_lossy());
+        let arguments = vec!["revert", "."];
+        let out = osc_command(&pkgpath, &arguments)?;
+        if out.status.success() {
+            let command_output = String::from_utf8_lossy(&out.stdout);
+            tracing::info!("✅ -- osc revert operation passed");
+            tracing::info!("stdout -- {}", command_output);
+        } else {
+            let command_output = String::from_utf8_lossy(&out.stderr);
+            tracing::info!("🚨 -- osc revert operation failed");
+            tracing::info!("stderr -- {}", command_output);
+        };
+        let arguments = vec!["clean", "."];
+        let out = osc_command(&pkgpath, &arguments)?;
+        if out.status.success() {
+            let command_output = String::from_utf8_lossy(&out.stdout);
+            tracing::info!("✅ -- osc clean operation passed");
+            tracing::info!("stdout -- {}", command_output);
+        } else {
+            let command_output = String::from_utf8_lossy(&out.stderr);
+            tracing::info!("🚨 -- osc clean operation failed");
+            tracing::info!("stderr -- {}", command_output);
+        };
```

Comment:
Let the code breath. Have spaces between "logical blocks". It's okay. That whitespace won't hurt, and helps people read and delineate things.

Reply:
i did somethings but cargo fmt be like 🔫